### DISCUSSION TITLE
Questions to other contibutors

### DISCUSSION THREAD

**Discussion by escamoteur** [escamoteur/web_gpu#8] [General]
Hi @jpnurmi, I only had a brief look at your added files and I haven't done anything with FFI yet, but I wonder why to we need Method channels at all?

**Comment by jpnurmi** [comment]
The WebGPU API is available in `libdawn_proc` (C), whereas the actual WebGPU implementation is in `libdawn_native` (C++). Or alternatively, `libdawn_wire` for sandboxed environments. Anyway, `libdawn_proc` is not aware of the native/wire implementations. The implementations provide "proc tables" that can be supplied to `libdawn_proc`. In practice, this means that we can utilize Dart FFI to resolve `wgpuFooBar` symbols from `libdawn_proc` in Dart, but we must manually initialize `libdawn_proc` to use `libdawn_native` in C++.

On desktop platforms, the initialization can be done from the platform plugin, and the readily available method channel seemed an easy way to invoke the initialization routine just to get started. In the long term, we should probably build our own `libdawn_init` helper library that provides the initialization routines and possibly the backend selection (OpenGL/Vulkan/D3D12/Metal) too. As soon as we have such a helper library, the method channel is probably no longer needed.

**Comment by escamoteur** [comment]
@jpnurmi did you have to generate the dawn sources witht this google build system, or how did you get the full sources?
Or did I missunderstand the whole description of the buildsystem?

**Comment by jpnurmi** [comment]
@escamoteur Yes, I built https://dawn.googlesource.com/dawn with the instructions in [building.md](https://dawn.googlesource.com/dawn/+/HEAD/docs/building.md) (gclient + gn + ninja). It worked out of the box for Linux and Windows. For Android it's probably easiest to build Dawn in the Chromium source tree.

**Reply by escamoteur** [reply]
Wow, that's amazing. I was a bit intimidated by that.
Can't we just take the sources and add them to the Android and iOS projects? I m not sure if tge chromium helps in any way as it's intended to be integrated into the browser.

**Reply by jpnurmi** [reply]
Do you mean building Dawn on the fly as demonstrated in https://medium.com/flutter-community/build-and-deploy-native-c-libraries-with-flutter-cc7531d590b5 ? I'm afraid Dawn is too big and complicated [*] to build like that. That approach only works for simpler libs that don't have large dependencies and build in a relatively short time.

[*] gclient sync pulls in several large dependencies, and the build needs depot tools, ninja etc. to begin with.

The reason why I'm looking into utilizing Chromium's source tree for building Dawn is that it seems to setup the dependencies correctly for Android, whereas the standalone Dawn source tree doesn't, unfortunately...

**Comment by escamoteur** [comment]
Ah, I see. are this big dependencies necessarry to build Dawn? But instead of on the fly couldn't it still be an option to manually add the sources to an adroid project, build them and then take tbe created binary lib?
Same problem are we facing with iOS/macos or not? Or is it possible to use the NDK directly as compiler?

**Comment by jpnurmi** [comment]
Bloating the repo with prebuilt libraries is suboptimal, but it has been such an adventure to get Dawn to build at all in the first place, that I'm not feeling too confident about extracting parts of the Dawn source tree and integrating them with the Flutter build system. Perhaps someone more familiar with Dawn/WebGPU and the GN build system would be up for the challenge... :smile:

Some interesting numbers for reference:
- `gclient sync` downloads ~2.2Gb (Linux)
  - includes e.g. i386 sysroot and probably plenty of other unnecessary suiff
- _full_ Linux x64 release build of Dawn takes ~4 minutes on i9-9980HK (8 cores, 16 threads)
  - includes unit tests and probably plenty of other unnecessary stuff

**Comment by escamoteur** [comment]
Yeah, I understand that. maybe the Rust version would be easier to handle.
How did you build the Android? And how will be build the iOS/MacOS ones?

**Comment by jpnurmi** [comment]
I haven't built Dawn for iOS yet, but I managed to build the macOS version yesterday. I built Dawn twice (`target_cpu = x86_64` and `target_cpu = arm64`) and used the `lipo` tool to create universal binaries.

**Comment by jpnurmi** [comment]
We can start collecting build instructions for each platform in the wiki once we have things figured out. I'm still unsure how to handle the initialization routine on macOS. I hope we can build and deploy the tiny web_gpu_init library on the fly, but I'm not an expert with CocoaPods so it may take some time to figure it out. :smile:
